rust/src/config/auth.rs

```diff
@@ -0,0 +1,134 @@
+use std::error::Error;
+use std::{borrow::Cow, fmt::Display};
+
+use nix::unistd::{Group, User};
+use serde::{de::Visitor, ser::SerializeTuple, Deserialize, Serialize};
+
+#[derive(Debug)]
+pub enum AuthEntrySerdeError {
+    UserNotFound(String),
+    GroupNotFound(String),
+    UserNotInGroup(String, String),
+}
+
+impl Display for AuthEntrySerdeError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            AuthEntrySerdeError::UserNotFound(user) => write!(f, "No user found for '{user}'"),
+            AuthEntrySerdeError::GroupNotFound(group) => write!(f, "No group found for '{group}'"),
+            AuthEntrySerdeError::UserNotInGroup(user, group) => {
+                write!(f, "User '{user}' is not part of group '{group}'")
+            }
+        }
+    }
+}
+
+impl Error for AuthEntrySerdeError {}
+
+/// A struct to deserialize user/group names
+/// into [`User`]/[`Group`].
+#[derive(Debug)]
+pub struct DummyAuthEntry {
+    pub user: User,
+    pub group: Group,
+}
+
+struct AuthEntryVisitor;
+
+impl<'de> Visitor<'de> for AuthEntryVisitor {
+    type Value = DummyAuthEntry;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("a tuple of user and group")
+    }
+
+    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+    where
+        A: serde::de::SeqAccess<'de>,
+    {
+        let user: Cow<str> = seq
+            .next_element()?
+            .ok_or_else(|| serde::de::Error::invalid_length(0, &self))?;
+        let user = User::from_name(&user)
+            .map_err(serde::de::Error::custom)?
+            .ok_or_else(|| {
+                serde::de::Error::custom(
+                    AuthEntrySerdeError::UserNotFound(user.to_string()).to_string(),
+                )
+            })?;
+
+        let group: Cow<str> = seq
+            .next_element()?
+            .ok_or_else(|| serde::de::Error::invalid_length(1, &self))?;
+        let group = Group::from_name(&group)
+            .map_err(serde::de::Error::custom)?
+            .ok_or_else(|| {
+                serde::de::Error::custom(
+                    AuthEntrySerdeError::GroupNotFound(group.to_string()).to_string(),
+                )
+            })?;
+
+        if user.gid != group.gid {
+            return Err(serde::de::Error::custom(
+                AuthEntrySerdeError::UserNotInGroup(user.name, group.name).to_string(),
+            ));
+        }
+
+        Ok(DummyAuthEntry { user, group })
+    }
+}
+
+impl<'de> Deserialize<'de> for DummyAuthEntry {
```

Comment:
Why can't you derive Serialize and Deserialize?

Reply:
Because `nix::User` and  `nix::Group` don't derive it themselves and I thought that it would be better to have the errors when deserializing the struct rather than build it after.  Do you think that there is a better way to do it?

Reply:
Hm, that is a problem.  I wonder if Nix should have a `serde` feature?  But that's a discussion for another time and place.